File: axi-py/packages/agenthub/agenthub/registry.py
"""Agent session registry — create, destroy, and look up sessions.

Module-level functions that take (hub, ...) as arguments.
Session dict (hub.sessions) is public — these functions mutate it directly.
"""

from __future__ import annotations

import logging
import os
from typing import TYPE_CHECKING, Any

from opentelemetry import trace

if TYPE_CHECKING:
    from agenthub.hub import AgentHub
    from agenthub.types import AgentSession

log = logging.getLogger(__name__)
_tracer = trace.get_tracer(__name__)


# ---------------------------------------------------------------------------
# Basic CRUD
# ---------------------------------------------------------------------------


def register_session(hub: AgentHub, session: AgentSession) -> None:
    """Add a session to the hub's registry."""
    hub.sessions[session.name] = session
    log.info("Session '%s' registered", session.name)


def unregister_session(hub: AgentHub, name: str) -> AgentSession | None:
    """Remove and return a session. Returns None if not found."""
    session = hub.sessions.pop(name, None)
    if session:
        log.info("Session '%s' unregistered", name)
    return session


def get_session(hub: AgentHub, name: str) -> AgentSession | None:
    """Look up a session by name."""
    return hub.sessions.get(name)


# ---------------------------------------------------------------------------
# End session
# ---------------------------------------------------------------------------


async def end_session(hub: AgentHub, name: str) -> None:
    """End a named session: disconnect client, release slot, remove from registry."""
    with _tracer.start_as_current_span("end_session", attributes={"agent.name": name}):
        session = hub.sessions.get(name)
        if session is None:
            return
        if session.client is not None:
            await hub.disconnect_client(session.client, name)
            session.client = None
            hub.scheduler.release_slot(name)
        if session.agent_log:
            for handler in list(session.agent_log.handlers):
                handler.close()
                session.agent_log.removeHandler(handler)
        hub.sessions.pop(name, None)
        log.info("Session '%s' ended", name)


# ---------------------------------------------------------------------------
# Rebuild / reset
# ---------------------------------------------------------------------------


async def rebuild_session(
    hub: AgentHub,
    name: str,
    *,
    cwd: str | None = None,
    session_id: str | None = None,
    system_prompt: Any = None,
    mcp_servers: dict[str, Any] | None = None,
) -> AgentSession:
    """End an existing session and create a fresh sleeping AgentSession.

    Preserves system prompt, cwd, and MCP servers from the old session
    unless overrides are provided.
    """
    from agenthub.types import AgentSession as SessionType

    old = hub.sessions.get(name)
    resolved_cwd = cwd or (old.cwd if old else "")
    resolved_prompt = system_prompt or (old.system_prompt if old else None)
    resolved_mcp = mcp_servers if mcp_servers is not None else (old.mcp_servers if old else None)
    old_frontend = old.frontend_state if old else None

    await end_session(hub, name)

    new_session = SessionType(
        name=name,
        cwd=resolved_cwd,
        system_prompt=resolved_prompt,
        client=None,
        session_id=session_id,
        mcp_servers=resolved_mcp,
        frontend_state=old_frontend,
    )
    hub.sessions[name] = new_session
    return new_session


async def reset_session(
    hub: AgentHub, name: str, cwd: str | None = None
) -> AgentSession:
    """Reset a named session. Preserves system prompt and MCP servers."""
    new_session = await rebuild_session(hub, name, cwd=cwd)
    log.info("Session '%s' reset (sleeping, cwd=%s)", name, new_session.cwd)
    return new_session


# ---------------------------------------------------------------------------
# Reclaim
# ---------------------------------------------------------------------------


async def reclaim_agent_name(hub: AgentHub, name: str) -> None:
    """If an agent with *name* already exists, kill it to free the name."""
    from agenthub import lifecycle

    if name not in hub.sessions:
        return
    _tracer.start_span(
        "reclaim_agent_name", attributes={"agent.name": name}
    ).end()
    log.info("Reclaiming agent name '%s' — terminating existing session", name)
    session = hub.sessions[name]
    await lifecycle.sleep_agent(hub, session, force=True)
    hub.sessions.pop(name, None)


# ---------------------------------------------------------------------------
# Spawn
# ---------------------------------------------------------------------------


async def spawn_agent(
    hub: AgentHub,
    *,
    name: str,
    cwd: str,
    agent_type: str = "claude_code",
    initial_prompt: str = "",
    resume: str | None = None,
    system_prompt: Any = None,
    mcp_servers: dict[str, Any] | None = None,
    frontend_state: Any = None,
) -> AgentSession:
    """Create a new agent session and register it.

    Does NOT run the initial prompt — caller is responsible for that
    (typically via messaging.run_initial_prompt or a frontend wrapper).
    This keeps spawn_agent a pure registry operation.
    """
    from agenthub.types import AgentSession as SessionType

    with _tracer.start_as_current_span(
        "spawn_agent",
        attributes={
            "agent.name": name,
            "agent.type": agent_type,
            "agent.cwd": cwd,
            "agent.resumed": bool(resume),
        },
    ):
        os.makedirs(cwd, exist_ok=True)

        session = SessionType(
            name=name,
            agent_type=agent_type,
            cwd=cwd,
            system_prompt=system_prompt,
            client=None,
            session_id=resume,
            mcp_servers=mcp_servers,
            frontend_state=frontend_state,
        )

        hub.sessions[name] = session
        log.info(
            "Agent '%s' registered (type=%s, cwd=%s, resume=%s)",
            name,
            agent_type,
            cwd,
            resume,
        )

        await hub.callbacks.on_spawn(session)
        return session

File: axi-rs/crates/axi/src/registry.rs
//! Agent session registry — create, destroy, and look up sessions.

use tracing::info;

use crate::lifecycle;
use crate::state::BotState;
use crate::types::AgentSession;

// ---------------------------------------------------------------------------
// Basic CRUD
// ---------------------------------------------------------------------------

pub async fn register_session(state: &BotState, session: AgentSession) {
    let name = session.name.clone();
    let mut sessions = state.sessions.lock().await;
    sessions.insert(name.clone(), session);
    info!("Session '{}' registered", name);
}

pub async fn unregister_session(state: &BotState, name: &str) -> Option<AgentSession> {
    let mut sessions = state.sessions.lock().await;
    let session = sessions.remove(name);
    if session.is_some() {
        info!("Session '{}' unregistered", name);
    }
    session
}

pub async fn get_session_names(state: &BotState) -> Vec<String> {
    let sessions = state.sessions.lock().await;
    sessions.keys().cloned().collect()
}

// ---------------------------------------------------------------------------
// End session
// ---------------------------------------------------------------------------

pub async fn end_session(state: &BotState, name: &str) {
    let has_client = {
        let sessions = state.sessions.lock().await;
        sessions.get(name).is_some_and(|s| s.awake)
    };

    if has_client {
        crate::claude_process::disconnect_client(state, name).await;
        let mut sessions = state.sessions.lock().await;
        if let Some(session) = sessions.get_mut(name) {
            session.awake = false;
        }
        state.scheduler().await.release_slot(name).await;
    }

    let mut sessions = state.sessions.lock().await;
    sessions.remove(name);
    info!("Session '{}' ended", name);
}

// ---------------------------------------------------------------------------
// Rebuild / reset
// ---------------------------------------------------------------------------

pub async fn rebuild_session(
    state: &BotState,
    name: &str,
    cwd: Option<String>,
    session_id: Option<String>,
    system_prompt: Option<serde_json::Value>,
    mcp_servers: Option<serde_json::Value>,
) -> AgentSession {
    let (old_cwd, old_prompt, old_mcp, old_sdk_mcp) = {
        let sessions = state.sessions.lock().await;
        sessions
            .get(name)
            .map(|old| {
                (
                    old.cwd.clone(),
                    old.system_prompt.clone(),
                    old.mcp_servers.clone(),
                    old.sdk_mcp_servers.clone(),
                )
            })
            .unwrap_or_default()
    };

    end_session(state, name).await;

    let mut new_session = AgentSession::new(name.to_string());
    new_session.cwd = cwd.unwrap_or(old_cwd);
    new_session.system_prompt = system_prompt.or(old_prompt);
    new_session.mcp_servers = if mcp_servers.is_some() {
        mcp_servers
    } else {
        old_mcp
    };
    new_session.session_id = session_id;
    new_session.sdk_mcp_servers = old_sdk_mcp;

    let mut sessions = state.sessions.lock().await;
    sessions.insert(name.to_string(), new_session);

    info!("Session '{}' rebuilt", name);

    drop(sessions);
    let sessions = state.sessions.lock().await;
    let mut result = AgentSession::new(name.to_string());
    if let Some(s) = sessions.get(name) {
        result.cwd = s.cwd.clone();
        result.system_prompt = s.system_prompt.clone();
        result.mcp_servers = s.mcp_servers.clone();
        result.session_id = s.session_id.clone();
        result.sdk_mcp_servers = s.sdk_mcp_servers.clone();
    }
    result
}

pub async fn reset_session(state: &BotState, name: &str, cwd: Option<String>) -> AgentSession {
    let new_session = rebuild_session(state, name, cwd.clone(), None, None, None).await;
    info!("Session '{}' reset (sleeping, cwd={:?})", name, cwd);
    new_session
}

// ---------------------------------------------------------------------------
// Reclaim
// ---------------------------------------------------------------------------

pub async fn reclaim_agent_name(state: &BotState, name: &str) {
    let exists = {
        let sessions = state.sessions.lock().await;
        sessions.contains_key(name)
    };
    if !exists {
        return;
    }
    info!(
        "Reclaiming agent name '{}' — terminating existing session",
        name
    );
    lifecycle::sleep_agent(state, name, true).await;
    let mut sessions = state.sessions.lock().await;
    sessions.remove(name);
}

// ---------------------------------------------------------------------------
// Spawn
// ---------------------------------------------------------------------------

/// Parameters for spawning a new agent session.
#[derive(Default)]
pub struct SpawnRequest {
    pub name: String,
    pub cwd: String,
    pub agent_type: Option<String>,
    pub resume: Option<String>,
    pub system_prompt: Option<serde_json::Value>,
    pub mcp_servers: Option<serde_json::Value>,
    pub mcp_server_names: Option<Vec<String>>,
}

pub async fn spawn_agent(state: &BotState, req: SpawnRequest) -> AgentSession {
    std::fs::create_dir_all(&req.cwd).ok();

    let name = req.name;
    let cwd = req.cwd;
    let resume = req.resume;

    let mut session = AgentSession::new(name.clone());
    session.agent_type = req.agent_type.unwrap_or_else(|| state.config.default_agent_type.clone());
    session.cwd = cwd.clone();
    session.system_prompt = req.system_prompt;
    session.session_id = resume.clone();
    session.mcp_servers = req.mcp_servers;
    session.mcp_server_names = req.mcp_server_names;

    {
        let mut sessions = state.sessions.lock().await;
        sessions.insert(name.clone(), session);
    }

    info!(
        "Agent '{}' registered (cwd={}, resume={:?})",
        name, cwd, resume
    );

    crate::frontend::on_spawn(state, &name).await;

    // Return a copy of session state
    let sessions = state.sessions.lock().await;
    let mut result = AgentSession::new(name.clone());
    if let Some(s) = sessions.get(&name) {
        result.cwd = s.cwd.clone();
        result.agent_type = s.agent_type.clone();
        result.session_id = s.session_id.clone();
        result.system_prompt = s.system_prompt.clone();
        result.mcp_servers = s.mcp_servers.clone();
        result.mcp_server_names = s.mcp_server_names.clone();
    }
    result
}

// ---------------------------------------------------------------------------
// Agent config persistence
// ---------------------------------------------------------------------------

/// Agent config stored alongside agent data.
#[derive(serde::Serialize, serde::Deserialize, Default)]
pub struct AgentConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mcp_server_names: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub packs: Option<Vec<String>>,
}

/// Save agent config to `<cwd>/agent_config.json`.
pub fn save_agent_config(cwd: &str, config: &AgentConfig) {
    let path = std::path::Path::new(cwd).join("agent_config.json");
    if let Ok(json) = serde_json::to_string_pretty(config) {
        if let Err(e) = std::fs::write(&path, json) {
            tracing::warn!("Failed to save agent config to {}: {}", path.display(), e);
        }
    }
}

/// Load agent config from `<cwd>/agent_config.json`.
pub fn load_agent_config(cwd: &str) -> Option<AgentConfig> {
    let path = std::path::Path::new(cwd).join("agent_config.json");
    let data = std::fs::read_to_string(&path).ok()?;
    serde_json::from_str(&data).ok()
}
